### PR TITLE
smb: cluster public ip addresses support

### PR DESCRIPTION
**Pull request by phlogistonjohn**
Depends on: <s>https://github.com/samba-in-kubernetes/sambacc/pull/133</s> pr merged

Add support for IPs managed by Samba's CTDB clustering system that can move due to cluster events. The IPs are configured in "interface form" (with a /<prefixlength>), and these values are passed down to ctdb through sambacc after being converted to device names by cephadm.

The cephadm service and the smb mgr module resource have basically the same YAML/JSON configuration. Example:

```yaml
- resource_type: ceph.smb.cluster
  cluster_id: adipctdb 
  auth_mode: active-directory
  domain_settings:
    realm: DOMAIN1.SINK.TEST
    join_sources:
      - source_type: resource
        ref: join1-admin
  # --> new stuff here:
  public_addrs:
    - address: 192.168.76.50/24
    - address: 192.168.76.51/24
    - address: 192.168.76.51/24
  # <--
  placement:
    count: 3
```

These can also take the form where a 2nd optional field, `destination` is provided with a network. this network is used by cephadm to get matched with a device name on the host.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), <s>opened tracker ticket</s>
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by phlogistonjohn** [comment]
jenkins test make check

**Comment by phlogistonjohn** [comment]
make check is struggling with protobuf stuff (again):
```
CMake Error at /usr/lib/x86_64-linux-gnu/cmake/protobuf/protobuf-config.cmake:14
```

**Comment by phlogistonjohn** [comment]
Known issue wrt jenkins: https://tracker.ceph.com/issues/64907

**Comment by phlogistonjohn** [comment]
jenkins test make check

**Comment by phlogistonjohn** [comment]
jenkins test make check

**Review by adk3798 — APPROVED** [review]

File: doc/cephadm/services/smb.rst [hunk 1]
```diff
@@ -105,21 +105,42 @@ custom_dns
...
 
-cluster_lock_uri:
+cluster_lock_uri
     A string containing a URI that identifies where Samba/CTDB will store a
     cluster lock. Required if ``clustered`` feature is set. Must be a RADOS
     pseudo-URI.
 
+cluster_public_addrs
+    List of objects; optional. Supported only when using Samba's clustering.
+    Assign "virtual" IPs that will be managed by the clustering subsystem
```
Comment (anthonyeleven) [inline doc/cephadm/services/smb.rst:125]: IP addresses

File: src/python-common/ceph/deployment/service_spec.py [hunk 2]
```diff
@@ -2842,6 +2842,126 @@ def validate(self) -> None:
...
+        except AttributeError:
+            return NotImplemented
+
+    def __repr__(self) -> str:
+        return (
+            f'SMBClusterPublicIPSpec({self.address!r}, {self.destination!r})'
+        )
+
+    def to_json(self) -> Dict[str, Any]:
+        """Return a json-compatible represenation of the SMBClusterPublicIPSpec."""
```
Comment (anthonyeleven) [inline src/python-common/ceph/deployment/service_spec.py:2916]: JSON
representation is misspelled

File: src/python-common/ceph/deployment/service_spec.py [hunk 3]
```diff
@@ -2842,6 +2842,126 @@ def validate(self) -> None:
...
+        out: Dict[str, Any] = {'address': self.address}
+        if self.destination:
+            out['destination'] = self.destination
+        return out
+
+    def to_strict(self) -> Dict[str, Any]:
+        """Return a strictly formed expanded json-compatible represenation of
```
Comment (anthonyeleven) [inline src/python-common/ceph/deployment/service_spec.py:2923]: Ditto

File: src/python-common/ceph/deployment/service_spec.py [hunk 4]
```diff
@@ -2842,6 +2842,126 @@ def validate(self) -> None:
...
+        the spec. This is not round-tripable.
```
Comment (anthonyeleven) [inline src/python-common/ceph/deployment/service_spec.py:2924]: Double p I should think

File: src/python-common/ceph/deployment/service_spec.py [hunk 5]
```diff
@@ -2896,6 +3016,10 @@ def __init__(
         # cluster_lock_uri - a pseudo-uri that resolves to a (rados) object
         # that will be used by CTDB for a cluster leader / recovery lock.
         cluster_lock_uri: Optional[str] = None,
+        # cluster_public_addrs - A list of SMB cluster public IP specs.
+        # If supplied, these will be used to esatablish floating virtual ips
```
Comment (anthonyeleven) [inline src/python-common/ceph/deployment/service_spec.py:3020]: IP addresses

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2024-08-27_12:51:47-orch:cephadm-wip-adk-testing-2024-08-26-1718-distro-default-smithi/

Failures:
- 6 `in cluster log` type failures, all of which are either known issue or not serious
- 1 `adjust-ulimits` command failing. Not something I'm familiar with but appears unrelated to orchestration
- 1 timeout in deploy raw OSDs test, seems unrelated to PRs in the run
- `test_create_export_via_apply` failure, caused by a specific PR in the run
- 2 `ceph orch ps`, `ceph versions` mismatch failures, known issue
- 2 mds_upgrade_sequence timeout failures, known issue
- 2 rgw-ingress timeout failures, known issue
- 2 dead jobs due to teuth/infra issues 

Only issue related to PRs in the run was the `test_create_export_via_apply` failure

File: doc/cephadm/services/smb.rst (hunk 1 above)
Comment (phlogistonjohn) [reply doc/cephadm/services/smb.rst]: done

File: src/python-common/ceph/deployment/service_spec.py (hunk 2 above)
Comment (phlogistonjohn) [reply src/python-common/ceph/deployment/service_spec.py]: done

File: src/python-common/ceph/deployment/service_spec.py (hunk 3 above)
Comment (phlogistonjohn) [reply src/python-common/ceph/deployment/service_spec.py]: done

File: src/python-common/ceph/deployment/service_spec.py (hunk 4 above)
Comment (phlogistonjohn) [reply src/python-common/ceph/deployment/service_spec.py]: changed

File: src/python-common/ceph/deployment/service_spec.py (hunk 5 above)
Comment (phlogistonjohn) [reply src/python-common/ceph/deployment/service_spec.py:3020]: done

**Review by obnoxxx — APPROVED** [review]
I do not claim to be an expert in cephadm, but what I read in the patch as documentation and code looks good from my Samba background. Thanks for adding this, @phlogistonjohn .   LGTM

**Comment by obnoxxx** [comment]
@phlogistonjohn one minor nit: how about calling it "CIDR form" instead of " interface form (with a /)"?

**Review by avanthakkar — APPROVED** [review]

**Comment by adk3798** [comment]
jenkins test docs

**Pull request merged**